File: year_2022/day_02/day02.py
from tools import loader, parsers

moves = {'A': 1, 'X': 1,  # rock
         'B': 2, 'Y': 2,  # paper
         'C': 3, 'Z': 3}  # scissors
wins = {(3, 1), (1, 2), (2, 3)}
res = {'X': 0, 'Y': 3, 'Z': 6}
test = ['A Y', 'B X', 'C Z']
reverse_wins = {1: {0: 3, 3: 1, 6: 2}, 2: {0: 1, 3: 2, 6: 3}, 3: {0: 2, 3: 3, 6: 1}}


def check_win(i: int, j: int) -> tuple[int, int]:
    if i == j:
        return 3, j
    if (i, j) in wins:
        return 6, j
    return 0, j


def part_1(data: list[str]) -> int:
    """test part 1:
    >>> print(part_1(test))
    15"""
    return sum(sum(check_win(moves[r[0]], moves[r[2]])) for r in data)


def part_2(data: list[str]) -> int:
    """test part 2:
    >>> print(part_2(test))
    12"""
    return sum(reverse_wins[moves[r[0]]][res[r[2]]] + res[r[2]] for r in data)


print(part_1(parsers.lines(loader.get())))  # 11603
print(part_2(parsers.lines(loader.get())))  # 12725

File: year_2022/day_02/day02.go
package main

import (
	"github.com/Nyaaa/advent-of-code/tools"
	"fmt"
	"strings"
)

func checkWin(left int, right int) int {
	if left == right {
		return 3
	} else if (left == 3 && right == 1) || (left == 1 && right == 2) || left == 2 && right == 3 {
		return 6
	}

	return 0
}

func play(data []string) (int, int) {
	moves := map[string]int{"A": 1, "X": 1, "B": 2, "Y": 2, "C": 3, "Z": 3}
	outcomes := map[string]int{"X": 0, "Y": 3, "Z": 6}
	part1 := 0
	part2 := 0

	for _, line := range data {
		if line != "" {
			letters := strings.Fields(line)
			left := moves[letters[0]]
			right := moves[letters[1]]
			part1 += right + checkWin(left, right)

			for i := 1; i <= 3; i++ {
				p2 := checkWin(left, i)
				if p2 == outcomes[letters[1]] {
					part2 += i + p2
				}
			}
		}
	}

	return part1, part2
}

func main() {
	testData := []string{"A Y", "B X", "C Z"}
	data := tools.ReadLines(tools.GetData(2022, 02))

	fmt.Println(play(testData))
	fmt.Println(play(data))
}
